plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/AuthorizationCodeReceiver.java

```diff
@@ -0,0 +1,139 @@
+/*******************************************************************************
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * All rights reserved. This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License v1.0 which
+ * accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ *******************************************************************************/
+
+package com.google.cloud.tools.eclipse.appengine.login;
+
+import com.google.common.base.Preconditions;
+
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.server.ServerConnector;
+import org.eclipse.jetty.server.handler.AbstractHandler;
+
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Runs a Jetty web server instance that opens up an HTTP listening port to receive OAuth2
+ * authorization code.
+ */
+public class AuthorizationCodeReceiver {
+
+  private static final Logger logger = Logger.getLogger(AuthorizationCodeReceiver.class.getName());
+
+  // TODO: create landing pages.
+  private static final String SUCCESS_LANDING_PAGE = "https://cloud.google.com/tools/intellij/auth_success";
+  private static final String FAILURE_LANDING_PAGE = "https://cloud.google.com/tools/intellij/auth_failure";
+
+  private static final String LOCAL_HOST = "127.0.0.1";
+  private static final String CODE_RECEPTION_POINT_PATH = "/code-reception-point";
+
+  private Server server;
+
+  private volatile boolean resultConsumed;
+
+  // Reads on these shared objects can happen only after server shutdown,
+  // so synchronization is not really necessary.
+  private String error;
+  private String authorizationCode;
+
+  /**
+   * Starts a local server listening through an ephemeral port.
+   *
+   * @return "http://127.0.0.1:<listening port>/code-reception-point", where this server listens
+   *     and expects to get an authorization code returning from a OAuth login flow
+   * @throws IOException wraps any {@link Exception} from Jetty when it failed to start
+   */
+  public String runLocalServer() throws IOException {
+    Preconditions.checkState(server == null, "Don't reuse the receiver.");
```

Comment:
Should be checkNull in this case, but there doesn't exist such a method, which makes sense because check methods throws an IllegalArgumentException. Here, I'm checking if the instance is in a proper state, so I think using checkState() makes sense, which throws IllegalStateException.